SYCL/ESIMD/api/functional/ctors/ctor_default.cpp

```diff
@@ -0,0 +1,171 @@
+              "default value, it value is: " +
+              std::to_string(it) +
+              ", but provided data type default value is: " +
+              std::to_string(default_val));
+        }
+      }
+    }
+
+    // verification for initializing simd by rvalue in an expression
+    {
+      std::vector<DataTypeT, AllocatorT> result{NumElems, AllocatorT{queue}};
+      queue.submit([&](sycl::handler &cgh) {
+        auto out = result.data();
+        cgh.single_task<kernel_for_simd<DataTypeT, 3, NumElems>>([=] {
+          simd<DataTypeT, NumElems> simd_by_rval;
+          simd_by_rval = simd<DataTypeT, NumElems>();
+          simd_by_rval.copy_to(out);
+        });
+      });
+      for (const auto &it : result) {
+        if (it != default_val) {
+          passed = false;
+          log.fail<DataTypeT, NumElems>(
+              "In simd by rvalue in an expression elem value is not equal to "
+              "default value, it value is: " +
+              std::to_string(it) +
+              ", but provided data type default value is: " +
+              std::to_string(default_val));
+        }
+      }
+    }
+
+    // verification for initializing simd by rvalue in actual parameter
+    {
+      std::vector<DataTypeT, AllocatorT> result{NumElems, AllocatorT{queue}};
+      queue.submit([&](sycl::handler &cgh) {
+        auto out = result.data();
+        auto obtain_simd_by_const_ref =
+            [=](const simd<DataTypeT, NumElems> &simd_by_const_ref) {
+              simd_by_const_ref.copy_to(out);
+            };
+        cgh.single_task<kernel_for_simd<DataTypeT, 4, NumElems>>(
+            [=] { obtain_simd_by_const_ref(simd<DataTypeT, NumElems>()); });
+      });
+      for (const auto &it : result) {
+        if (it != default_val) {
+          passed = false;
+          log.fail<DataTypeT, NumElems>(
+              "In simd by const reference elem value is not equal to default "
+              "value, it value is: " +
+              std::to_string(it) +
+              ", but provided data type default value is: " +
+              std::to_string(DataTypeT{}));
+        }
+      } // cycle "for"
+    }
+  }
+};
+
+template <typename DataTypeT> using run_test_with_one_elem = test<DataTypeT, 1>;
+
+template <typename DataTypeT>
+using run_test_with_eight_elems = test<DataTypeT, 8>;
+
+template <typename DataTypeT>
+using run_test_with_sixteen_elems = test<DataTypeT, 16>;
+
+template <typename DataTypeT>
+using run_test_with_thirty_two_elems = test<DataTypeT, 32>;
+
+int main(int argc, char **argv) {
+  logger::Logger logger{};
+  sycl::queue queue{};
+
+  bool passed{true};
+
+  for_all_types<run_test_with_one_elem>(types, logger, queue, passed);
+  for_all_types<run_test_with_eight_elems>(types, logger, queue, passed);
+  for_all_types<run_test_with_sixteen_elems>(types, logger, queue, passed);
+  for_all_types<run_test_with_thirty_two_elems>(types, logger, queue, passed);
```

Comment:
```suggestion
  for_all_types<test, 1>(types, logger, queue, passed);
  for_all_types<test, 8>(types, logger, queue, passed);
  for_all_types<test, 16>(types, logger, queue, passed);
  for_all_types<test, 32>(types, logger, queue, passed);
```
Why not something like this?

Reply:
`for_all_types` can't obtain only type arguments